### PR TITLE
Feature 5 refactoring

### PR DESCRIPTION
**Pull request by orhanna14**
This PR does a few things, and has a few questions: 
1) It adds a `ComputerPlayer` class and renames the `PlayerInput` class to `Player`. Both classes have a more abstract method of `take_turn`. (We had talked a little about polymorphism, and I experimented with it a little on the side, _is this the time to definitely put that in?_ If so, why now?)

2) It develops the `Grid` class a bit. The logic is extracted for `player x` and `player o`, there is a method added for valid computer options, and it experiments with the `structure` of the grid to be a collection of coordinates... which is then assigned to keys in a hash. _(I'm not sure why/if this is better?)_

### CODE REVIEW

File: lib/computer_player.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class ComputerPlayer
```
Comment (baccigalupi) [inline lib/computer_player.rb:1]: I like the symmetry/polymorphism you created with a computer player and a player!

File: lib/computer_player.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  attr_accessor :computer_choices, :grid
+
+  def initialize(grid)
+    @computer_choices = []
```
Comment (baccigalupi) [inline lib/computer_player.rb:5]: These `computer_choices` was hard to follow. From what I can guess having taken a couple passes, the grid returns all valid choices and then it is stored.

While this class only has `take_turn` as a public method so the risk of stale data is not present. It is weird that the class holds on to this data between turns.

If we think about the state as being essential to the object, because objects are combinations of state with methods, then that it exists in this out of sync world where the available computer choices become wrong as soon as the other player takes a turn ... it feels wrong.

Could we just ask the grid for a single random option? If you want the computer to have the option to make its own choice without the grid being involved. What if this computer player had its own grid, independent from the shared game grid. In that way it could ask for options that are already bounded and valid.

This would mean that the class would become closer to the input player class because it would loop to ask if a coordinate was valid.

File: lib/game_runner.rb [hunk 3]
```diff
@@ -20,13 +21,18 @@ def grid_printer
...
+  end
+
+  def computer_player
+    @computer_player ||= ComputerPlayer.new(grid)
   end
 
   def run
     introduce_game
-    player_1_turn
+    player_turn
```
Comment (baccigalupi) [inline lib/game_runner.rb:34]: This is very cool that it boils down to two similar statements. 

What would it take to have it be this instead

```ruby
player.take_turn
computer_player.take_turn
```

A lot of stuff would need to move to those player classes. Also, I wonder about the players maintaining their own markers!

File: lib/grid.rb [hunk 4]
```diff
@@ -2,11 +2,27 @@ class Grid
   attr_accessor :structure
 
   def initialize
-    @structure = grid_structure
+    @structure = build_grid
   end
 
-  def mark(coordinate)
-    structure[coordinate] = 'X'
+  def player_x
```
Comment (baccigalupi) [inline lib/grid.rb:8]: I think before we talked about moving these markers here so that random objects wouldn't have to know about the marker. But now that you have players, it makes a lot of sense to me that they would know their own marker. If we wanted to have emoji characters for player marks instead of X and O, we could.

File: lib/grid.rb [hunk 5]
```diff
@@ -17,9 +33,31 @@ def get_value(coordinate)
...
-  def grid_structure
-    {"A1"=> ' ', "A2"=> ' ', "A3"=> ' ', "B1"=> ' ', "B2"=> ' ', "B3"=> ' ', "C1"=> ' ', "C2"=> ' ', "C3"=> ' '}
+  def build_grid
+    coordinates = []
+    row.each do |letter|
+      column.each do |number|
+         coordinates.push(letter + number)
+      end
+    end
+    Hash[coordinates.collect { |coordinate| [coordinate, ' '] } ]
```
Comment (baccigalupi) [inline lib/grid.rb:53]: I'm wondering how this programmatic build up of the grid is better than the literal? It would be useful if we were going to dynamically change the size of the grid, but that isn't this feature.

I know we had originally talked about changing the structure of the grid, but the goal was to shove data into the `Cell` object. It would have data: `coordinate` and `mark`. The coordinates would be the string row and column; the mark would be X, O or nil.

File: lib/grid.rb [hunk 6]
```diff
@@ -17,9 +33,31 @@ def get_value(coordinate)
     structure[coordinate]
   end
 
+  def get_valid_options(computer_choices)
+    structure.each { |key, value|
```
Comment (baccigalupi) [inline lib/grid.rb:37]: Indentation off here in this block.

File: spec/computer_player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
+require "spec_helper"
+require_relative "../lib/computer_player.rb"
+require_relative "../lib/grid.rb"
+
+RSpec.describe ComputerPlayer do
+  describe "#take_turn" do
+    it "gives the computer all choices that are available" do
```
Comment (baccigalupi) [inline spec/computer_player_spec.rb:7]: This test is a bit hard to follow. I think because the functionality is hard to follow too.

File: spec/game_runner_spec.rb [hunk 8]
```diff
@@ -60,7 +60,7 @@
       
       it "after the user takes a turn, the computer takes a turn, and the grid is re-rendered with an O in a random location" do
         def seed_random_generator_for_computer_turn
```
Comment (baccigalupi) [inline spec/game_runner_spec.rb:62]: I didn't notice this method in the middle of the it block before. It should probably move into a support file, or at least into the highest describe block.

**Review by baccigalupi — CHANGES_REQUESTED** [review]
Getting really interesting. Let me know if you want to get on a call. I know it has been a while.

File: lib/grid.rb (hunk 4 above)
Comment (orhanna14) [reply lib/grid.rb:8]: ✔️

File: lib/grid.rb (hunk 6 above)
Comment (orhanna14) [reply lib/grid.rb:37]: TY!

File: spec/game_runner_spec.rb (hunk 8 above)
Comment (orhanna14) [reply spec/game_runner_spec.rb:62]: Right. ✔️
